### PR TITLE
#258 refactor flux implementation

### LINKED ISSUE #258: frontend refactorings
## Description

The frontend still contains some design flaws that must be fixed:
- The `flux-container` is ready to use and should be introduced into Sententiaregum before continuing with #35 as this story requires an enhanced flux implementation like our newly created on
- bump dependencies (especially react-bootstrap, react and jsdom)
- use node6 in travis-ci to have proper builds
- get rid of the `TestUtils`
- rewrite the page structure (traverse over certain components and compose them with each other instead of leaky inheritance concepts)
- find an approach to simplify the form markup

### PR DESCRIPTION
**Pull request by Ma27**
### Overview

fixes certain mistakes in the frontend.
The following descisions have been reverted;
- `flux` as only tool to implement the flux pattern
- abstract component API (caused re-renderings of the menu bar and the whole container, new approach re-renders the content and keeps the menu)

**Done:**
- [x] use `edge` image rather than precise and NodeJS 6 in the build environment
- [x] major dependency bump (e.g. ReactJS 15, react-router 2.0, JSDom 8)
- [x] simplify service structure in `util/http`
- [x] lightweight form API to reduce boilerplate
- [x] structure refactoring to use containers rather than abstract classes to generate boiler-plate JSX (fixes issue with loclae initialization)
- [x] use flux-container to reduce boilerplate in the business logic
- [x] minor refactorings
### Related tickets

resolves #258

### CODE REVIEW

**Comment by Ma27** [comment]
before merging, the commits must be squashed in chunks (too much right now, but squash isn't a good idea)

File: .gitignore [hunk 1]
```diff
@@ -78,3 +78,6 @@ npm-debug.log
 
 # Capistrano Configuration
 /deploy/config/deploy.yaml
+
+# LocalStorage Emulation
+./scratch/*
```
Comment (Ma27) [inline .gitignore]: I admit it, I've copied it from the docs. The dir must be renamed of course.

File: src/AppBundle/Controller/BaseController.php [hunk 2]
```diff
@@ -62,4 +88,24 @@ protected function getCurrentUser()
...
+    {
+        return $this->getParameter('app.locales');
+    }
+
+    /**
+     * Getter for the locale keys.
+     *
+     * @return string[]
+     */
+    protected function getLocaleKeys()
```
Comment (Ma27) [inline src/AppBundle/Controller/BaseController.php]: how about _getLocaleShortNames_?

File: src/AppBundle/View/I18nResponseBuilder.php [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+ */
+
+namespace AppBundle\View;
+
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\Validator\ConstraintViolation;
+use Symfony\Component\Validator\ConstraintViolationListInterface;
+
+/**
+ * Concrete implementation of a builder which transforms constraint violations into an i18n response.
```
Comment (Ma27) [inline src/AppBundle/View/I18nResponseBuilder.php]: it doesn't build the response, but the format for it

File: src/AppBundle/View/I18nResponseBuilder.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+ *   'another_property' => [
+ *     [
+ *       // ...
+ *     ]
+ *   ]
+ * ]
+ *
+ * @author Maximilian Bosch <maximilian.bosch.27@gmail.com>
+ */
+class I18nResponseBuilder implements I18nResponseBuilderInterface
```
Comment (Ma27) [inline src/AppBundle/View/I18nResponseBuilder.php]: rename to _I18nResponseFormatBuilder_ (see comment above)?

File: src/AppBundle/View/I18nResponseBuilder.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+     */
+    public function __construct(TranslatorInterface $translator)
+    {
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildTranslatableResponse(
```
Comment (Ma27) [inline src/AppBundle/View/I18nResponseBuilder.php]: rename to _formatTranslatableViolationList_?

File: src/AppBundle/Controller/BaseController.php (hunk 2 above)
Comment (benbieler) [inline src/AppBundle/Controller/BaseController.php]: Yes, could be usefull...

**Comment by Ma27** [comment]
@benbieler have to finish my review, but would you approve?

File: package.json [hunk 6]
```diff
@@ -35,30 +36,31 @@
...
     "npmconf": "~2.1.2",
     "object.assign": "^4.0.3",
     "perfect-scrollbar": "^0.6.11",
-    "react": "^0.14",
-    "react-bootstrap": "^0.28",
-    "react-dom": "^0.14",
-    "react-interpolate-component": "^0.8",
+    "react": "^15.0.2",
+    "react-bootstrap": "^0.29",
+    "react-dom": "^15.0",
```
Comment (Ma27) [inline package.json:53]: should be increased to `^15.0.1` as 15.0 contained a regression bug

File: src/AppBundle/View/I18nResponseBuilder.php [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        $violationList = iterator_to_array($violations, false);
+        $hasLocales    = count($targetLocales) > 0;
+
+        if (!$useAllLanguages && $hasLocales || $useAllLanguages && !$hasLocales) {
+            throw new \InvalidArgumentException(
+                'Wrong usage of $targetLocales: If all languages should be rendered, $targetLocales should be set, if not, $targetLocales should be an empty array!'
+            );
+        }
+
+        return array_reduce(
```
Comment (Ma27) [inline src/AppBundle/View/I18nResponseBuilder.php]: I'm still not sure whether `array_walk` is better or an imperative foreach approach

File: src/AppBundle/View/I18nResponseBuilder.php [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+                    //
+                    // [
+                    //   'property' => [
+                    //     // this is the structure merged into the property.
+                    //     'de' => ['Deutscher Text'],
+                    //     'en' => ['English text'],
+                    //   ]
+                    // ]
+                    case $useAllLanguages:
+                        foreach ($targetLocales as $locale) {
```
Comment (Ma27) [inline src/AppBundle/View/I18nResponseBuilder.php]: another point where we should evaluate whether the imperative approach or a `array_` function is the better solution

File: src/Frontend/config/counterpart.js [hunk 9]
```diff
@@ -10,15 +10,11 @@
...
 
-import { jsdom } from 'jsdom';
+import de from './languages/de';
+import en from './languages/en';
 import counterpart from 'counterpart';
-import de from '../config/languages/de';
-import en from '../config/languages/en';
+import Locale from '../util/http/LocaleService';
 
+Locale.setLocale(null);
```
Comment (Ma27) [inline src/Frontend/config/counterpart.js:21]: requires a comment as setting the locale to `null` may seem a little bit odd during debug

File: src/Frontend/store/handler/filterItemsByVisibility.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+ * Custom store handler which filters items by some auth configuration.
+ *
+ * @param {Array}  items    The menu items.
+ * @param {Object} authData Information about the authentication status.
+ *
+ * @returns {Array} The new state.
+ */
+export default (items, authData) => {
+  return items.filter(item => {
+    return !('ROLE_ADMIN' === item.role && !authData.is_admin
```
Comment (Ma27) [inline src/Frontend/store/handler/filterItemsByVisibility.js]: the first condition should be in a new line, too.

**Comment by Ma27** [comment]
except the failing style validation and the comments, it can be chunked properly and finally merged

**Comment by Ma27** [comment]
everythign seems fine now.
Some commits need squash, certain things should be splitted and in the end we have to reorder stuff a bit

**Comment by Ma27** [comment]
ready for now. I've squashed most of the commits and moved the tests and review fixes to the correct commits to avoid those `fixes review comments` commits and keep the history clean.

File: src/Frontend/components/app/layout/Application.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * Container for the whole app.
+ *
+ * @param {Object} props The properties of the app container.
+ *
+ * @returns {React.Element} The markup of the app.
+ */
+const Application = props => {
+  return (
```
Comment (Ma27) [inline src/Frontend/components/app/layout/Application.js:25]: **NOTE:** I didn't implement a refresh menu feature which re-renders the menu structure if something changes. Instead it should listen to the login/logout lifecycle to be implemented in #35 and decide based on the given data of the login/logout cycle whether to re-render. This saves unncessary resources that would be needed to re-evaluate the menu items at every hash-change

**Pull request merged**